kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/Coordinator.java

```diff
@@ -255,6 +305,127 @@ private void commitToTable(
     }
   }
 
+  private static class PartitionOffset {
+
+    private final int partition;
+    private final long offset;
+
+    PartitionOffset(int partition, long offset) {
+
+      this.partition = partition;
+      this.offset = offset;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      PartitionOffset that = (PartitionOffset) o;
+      return partition == that.partition && offset == that.offset;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(partition, offset);
+    }
+
+    @Override
+    public String toString() {
+      return "PartitionOffset{" + "partition=" + partition + ", offset=" + offset + '}';
+    }
+  }
+
+  private List<Envelope> deduplicateEnvelopes(
+      List<Envelope> envelopes, UUID commitId, TableIdentifier tableIdentifier) {
+    return deduplicate(
+        envelopes,
+        envelope -> new PartitionOffset(envelope.partition(), envelope.offset()),
+        (numDuplicatedFiles, duplicatedPartitionOffset) ->
+            String.format(
+                "Detected %d envelopes with the same partition-offset=%s during commitId=%s for table=%s",
```

Comment:
If we see this log message, it means the coordinator has somehow read the same message (same partition-offset) twice.

Reply:
No unit test for this message unfortunately, the `MockConsumer` class is too smart (i can't add records with the same partition-offset :D)

Reply:
Exactly, so solve it by pulling this out to a class as a pure function?

Reply:
Fixed, and there are tests for it now :)